File: cart/src/main/java/com/github/posko/cart/domain/model/SelectedProduct.java
package com.github.posko.cart.domain.model;

import android.support.annotation.NonNull;

public class SelectedProduct {

    private Integer productId;

    private Integer variantId;

    private Integer unitId;

    private Double price;

    private Integer quantity;

    public Integer getProductId() {
        return productId;
    }

    public Integer getVariantId() {
        return variantId;
    }

    public Integer getUnitId() {
        return unitId;
    }

    public Double getPrice() {
        return price;
    }

    public Integer getQuantity() {
        return quantity;
    }

    public Double getTotal() {
        return price * quantity;
    }

    public SelectedProduct(Builder builder) {
        this.productId  = builder.productId;
        this.variantId  = builder.variantId;
        this.unitId     = builder.unitId;
        this.price      = builder.price;
        this.quantity   = builder.quantity;
    }

    public static class Builder {

        private Integer productId;

        private Integer variantId;

        private Integer unitId;

        private Double price;

        private Integer quantity;


        public Builder setProductId(@NonNull Integer productId) {
            this.productId = productId;
            return this;
        }

        public Builder setVariantId(@NonNull Integer variantId) {
            this.variantId = variantId;
            return this;
        }

        public Builder setUnitId(@NonNull Integer unitId) {
            this.unitId = unitId;
            return this;
        }

        public Builder setPrice(@NonNull Double price) {
            this.price = price;
            return this;
        }

        public Builder setQuantity(@NonNull Integer quantity) {
            this.quantity = quantity;
            return this;
        }

        public SelectedProduct build() {
            return new SelectedProduct(this);
        }
    }
}

File: cart/src/main/java/com/github/posko/cart/domain/model/SelectedProduct.kt
package com.github.posko.cart.domain.model

class SelectedProduct private constructor(builder: Builder) {

    private val productId: Int?

    private val variantId: Int?

    private val unitId: Int?

    private val price: Double?

    private val quantity: Int?

    private val total: Double
        get() = quantity!! * price!!

    init {
        this.productId  = builder.productId
        this.variantId  = builder.variantId
        this.unitId     = builder.unitId
        this.price      = builder.price
        this.quantity   = builder.quantity
    }

    class Builder {

        var productId: Int? = null

        var variantId: Int? = null

        var unitId: Int? = null

        var price: Double? = null

        var quantity: Int? = null


        fun setProductId(productId: Int?): Builder {
            this.productId = productId
            return this
        }

        fun setVariantId(variantId: Int?): Builder {
            this.variantId = variantId
            return this
        }

        fun setUnitId(unitId: Int?): Builder {
            this.unitId = unitId
            return this
        }

        fun setPrice(price: Double?): Builder {
            this.price = price
            return this
        }

        fun setQuantity(quantity: Int?): Builder {
            this.quantity = quantity
            return this
        }

        fun create(): SelectedProduct {
            return SelectedProduct(this)
        }
    }

}
